autosklearn/ensemble_building/builder.py

```diff
@@ -0,0 +1,925 @@
+        pynisher_context: str | None = None,
+        time_left: float | None = None,
+        end_at: float | None = None,
+        time_buffer: int = 5,
+    ) -> tuple[list[dict[str, Any]], int | float]:
+        """Run the ensemble building process
+
+        Parameters
+        ----------
+        iteration : int
+            What iteration to associate with this run
+
+        pynisher_context : str | None = None
+            The pynisher context to run in. If None, defaults to
+            multiprocessing.get_context(None)
+
+        time_left : float | None = None
+            How much time should be left for this run. Either this or `end_at` must
+            be provided.
+
+        end_at : float | None = Non
+            When this run should end. Either this or `time_left` must be provided.
+
+        time_buffer : int = 5
+            How much extra time to add as a buffer to this run. This means there is
+            always some amount of time to do something useful.
+
+        Returns
+        -------
+        (ensemble_history, nbest)
+        """
+        if time_left is None and end_at is None:
+            raise ValueError("Must provide either time_left or end_at.")
+
+        elif time_left is not None and end_at is not None:
+            raise ValueError("Cannot provide both time_left and end_at.")
+
+        if not self.logger:
+            self.logger = get_named_client_logger(
+                name="EnsembleBuilder",
+                port=self.logger_port,
+            )
+
+        process_start_time = time.time()
+        while True:
+
+            if time_left is not None:
+                time_elapsed = time.time() - process_start_time
+                time_left -= time_elapsed
+            else:
+                assert end_at is not None
+                current_time = time.time()
+                if current_time > end_at:
+                    break
+                else:
+                    time_left = end_at - current_time
+
+            wall_time_in_s = int(time_left - time_buffer)
+            if wall_time_in_s < 1:
+                break
+
+            context = multiprocessing.get_context(pynisher_context)
+            preload_modules(context)
+
+            safe_ensemble_script = pynisher.enforce_limits(
+                wall_time_in_s=wall_time_in_s,
+                mem_in_mb=self.memory_limit,
+                logger=self.logger,
+                context=context,
+            )(self.main)
+
+            safe_ensemble_script(time_left, iteration)
+
+            status = safe_ensemble_script.exit_status
+            if isinstance(status, pynisher.MemorylimitException):
+                # if ensemble script died because of memory error,
+                # reduce nbest to reduce memory consumption and try it again
+
+                # ATTENTION: main will start from scratch;
+                # all data structures are empty again
+                try:
+                    self.previous_candidates_path.unlink()
```

Comment:
Trying to migrate things over to Pathlib more for which this is synonymous. See the [docs](https://docs.python.org/3/library/pathlib.html#correspondence-to-tools-in-the-os-module). Basically there is no method `remove` in pathlib but this is the same thing.

The reason to move over to pathlib more is that it would gives users much more ability to do things, i.e. have a path to a remote server of some kind. `os` does not allow for anything like that.

Reply:
Ah, sorry, my bad. From my brief googling I though pathlib has a remove function, too.